### PR TITLE
Check server state when bring nodes back

### PR DESCRIPTION
**Pull request by SidW**
Fixed the issue when starting up Voldemort in offilne state, it actually
goes online and keeps listening client requests.

### CODE REVIEW

File: src/java/voldemort/server/VoldemortServer.java [hunk 1]
```diff
@@ -382,7 +382,15 @@ protected void startInner() throws VoldemortException {
...
         long start = System.currentTimeMillis();
-        boolean goOnline = true;
+
+        boolean goOnline;
+        if (getMetadataStore().getServerStateUnlocked() == MetadataStore.VoldemortState.NORMAL_SERVER)
+            goOnline = true;
+        else if (getMetadataStore().getServerStateUnlocked() == MetadataStore.VoldemortState.OFFLINE_SERVER)
+            goOnline = false;
+        else
+            throw new VoldemortException("Invalid Server state. Neither online nor offline found");
```
Comment (arunthirupathi) [inline src/java/voldemort/server/VoldemortServer.java]: The server has 3 states, online, offline and rebalancing. This makes the server not to start when that is in the Rebalancing mode. I would like to have only 2 cases here, offline and everything else.

File: test/unit/voldemort/client/OfflineStateTest.java [hunk 2]
```diff
@@ -262,4 +268,23 @@ public void testStateTransitions() throws InterruptedException {
...
+
+        String baseDirPath = server.getVoldemortConfig().getVoldemortHome();
+        String parentDirPath = Paths.get(baseDirPath).getParent().toString();
+        List<StoreDefinition> storeDefs = server.getMetadataStore().getStoreDefList();
+        try {
+            VoldemortConfig config =
+                    ServerTestUtils.createServerConfigWithDefs(useNio, 0, parentDirPath, cluster, storeDefs, new Properties());
+            VoldemortServer newServer = ServerTestUtils.startVoldemortServer(socketStoreFactory, config, cluster);
+            fail("should have thrown an connection exception");
+        }catch (VoldemortException e){
```
Comment (arunthirupathi) [inline test/unit/voldemort/client/OfflineStateTest.java]: I would just handle Catch UnreachableStoreException and also check that the inner exception is ConnectException.class.

File: test/unit/voldemort/client/OfflineStateTest.java [hunk 3]
```diff
@@ -262,4 +268,23 @@ public void testStateTransitions() throws InterruptedException {
...
+            assertEquals(UnreachableStoreException.class, e.getClass());
+        }
+    }
```
Comment (arunthirupathi) [inline test/unit/voldemort/client/OfflineStateTest.java]: I would also like to add that the server that was brought up in offline mode can go to online mode. by calling the appropriate methods.

File: test/unit/voldemort/client/OfflineStateTest.java [hunk 4]
```diff
@@ -262,4 +268,23 @@ public void testStateTransitions() throws InterruptedException {
         assertTrue(testOnlineTraffic());
         assertTrue(testSlopStreaming());
     }
+
+    @Test
```
Comment (arunthirupathi) [inline test/unit/voldemort/client/OfflineStateTest.java]: I would also like to add one more test, where a server stopped in online mode, can be brought up in offline mode. The steps that are done in the test, need to be documented and given to SREs.

File: src/java/voldemort/server/VoldemortServer.java [hunk 5]
```diff
@@ -386,10 +386,8 @@ protected void startInner() throws VoldemortException {
         boolean goOnline;
         if (getMetadataStore().getServerStateUnlocked() == MetadataStore.VoldemortState.NORMAL_SERVER)
```
Comment (arunthirupathi) [inline src/java/voldemort/server/VoldemortServer.java]: Rebalance Server should still come as online. This would make the rebalance server to go offline.

**Comment by arunthirupathi** [comment]
The test looks good now. If you can make minor tweak to the original logic and squash your commit, I will merge the commit.

**Comment by SidW** [comment]
Thanks Arun! I should have been more careful. I have modified that and
rebased my commits. Ready to be merged😀

**Comment by arunthirupathi** [comment]
Rebased and merged  https://github.com/voldemort/voldemort/commit/af3d605f9b953f1caa0bd36cde3f1e3a0635da3e

**Pull request closed without merging**